File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.kt
class Solution {
    fun solution(players: Array<String>, callings: Array<String>): Array<String> {
        val playersList = players.toMutableList()
        var rank=0; val map=players.map {it to rank++  }.toMap().toMutableMap()
        callings.forEach { calledPlayer ->
            val currentIndex = map[calledPlayer]!!
            if (currentIndex > 0) {
                val previousPlayer = playersList[currentIndex - 1]
                playersList[currentIndex - 1] = calledPlayer
                playersList[currentIndex] = previousPlayer
                map[calledPlayer] = currentIndex - 1
                map[previousPlayer] = currentIndex
            }
        }
        return playersList.toTypedArray()
    }
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_map>
#include <algorithm>
using namespace std;

vector<string> solution(vector<string> players, vector<string> callings) {
    unordered_map<string,int>map;
    for (int i = 0; i < players.size(); i++)
    {
        map.insert(make_pair(players[i], i));
    }
    for (string str : callings)
    {
        string sub_str = players[map[str] - 1];
        swap(players[map[str]], players[map[sub_str]]);
        swap(map[str], map[sub_str]);
    }
    return players;
}
